# Практическая работа №2: Обработка выборочных данных. Нахождение точечных оценок параметров распределения

Выполнили студентки гр. 2383 Щёголева Надежда и Семененко Анна. Вариант №28

## Цель работы

Получение практических навыков нахождения точечных статистических оценок параметров распределения.

## Основные теоретические положения
### Начальный эмпирический момент
Начальным эмпирическим моментом к-го порядка называется среднее значение к-х степеней элементов вариационного ряда:
$$\bar{M}_k = \frac{1}{N} \sum n_jx_j^k$$

### Центральный эмпирический момент
Центральным эмпирическим моментом к-го порядка называется среднее значение к-х степеней разностей $(x_j - \bar{x}_в)$ для вариационного или интервального ряда:
$$\bar{m}_k = \cfrac{1}{N}\sum n_j(x_j - \bar{x}_в)^k$$

### Статистическая оченка дисперсии(смещенная):
$$D_в = \bar{m_2} = \cfrac{1}{N}\sum n_j(x_j - \bar{x}_в)^2$$

### Исправленная оценка дисперсии:
$$s^2 = \cfrac{N}{N-1}D_в$$

### Статистические оценки СКО:
$$\sigma_в = \sqrt{D_в};   s = \sqrt{s^2}$$

### Статистические оценки ассиметрии и эксцесса
$$\bar{A}_s = \cfrac{\bar{m_3}}{s^3}; \bar{E} = \cfrac{\bar{m_4}}{s^4} - 3$$

Если $\bar{E} >0$, то эмпирическое распределение является более высоким (островершинным) – относительно нормального распределения. Если же $\bar{E} < 0 $– то более низким и пологим. И чем больше $\bar{E}$ по модулю, тем аномальнее высота в ту или иную сторону.

### Условные варианты
$$u_j = \cfrac{x_j -C}{h},$$

где С - условный ноль, значение которого выбирается равным значению варинаты интервального ряда, являющейся средней или близкой к средней по значению в этом ряду. В результате все условные варианты оказываются целыми числами.

### Условные моменты к-того порядка:
$$\bar{M^*_k} = \cfrac{1}{N} \sum n_j \bigg(\cfrac{x_j -C}{h}\bigg)^k = \cfrac{1}{N}\sum n_ju_j^k$$

### Мода

Мода $M^*_o$ вариационного ряда - это варианта с наибольшей частотой.

Мода интервального ряда находится для интервала $(x_{i-1}, x_i)$:

$$M_o^* = x_{i-1} + h \cdot \frac{\tilde{m}_i - \tilde{m}_{i-1}}{(\tilde{m}_i - \tilde{m}_{i-1}) + (\tilde{m}_i - \tilde{m}_{i+1})}$$

где:
- $x_{i-1}$ — начало модального интервала;
- $h$ — ширина интервала;
- $\tilde{m}_i$ — относительная частота модального интервала;
- $\tilde{m}_{i-1}$, $\tilde{m}_{i+1}$ — относительные частоты соседних интервалов.

### Медиана
Медиана $M_e^*$ вариационного ряда — это варианта, соответствующая середине ранжированного ряда.

Для ряда с нечетным числом элементов

$$M_e^* = x_{m+1}$$

Для ряда с четным числом элементов

$$M_e^* = \frac{x_m + x_{m+1}}{2}$$

Для интервального ряда выборочное значение медианы может быть вычислено по формуле:

$$M_e^* = x_{i-1} + h \cdot \frac{0.5 - \sum_{j=1}^{i-1} \tilde{m}_j}{\tilde{m}_i}$$

где:
* $x_{i-1}$ — начало медианного интервала;
* $m_i$ — относительная частота медианного интервала;
* $\sum_{j=1}^{i-1} m_j$ — накопленная относительная частота для интервалов, предшествующих медианному.

Условие выбора медианного интервала:

Значение медианы принадлежит интервалу, для которого накопленная относительная частота равна или превышает 0.5, а накопленная относительная частота для предыдущего интервала меньше 0.5.

Прямая $x = M_e^*$ делит площадь гистограммы на две равные части.

### Выборочный коэффициент вариации

Выборочный коэффициент вариации $V^*$ характеризует относительный разброс значений вариант относительно выборочного среднего и выражается в процентах:

$$V^* = \frac{\sigma_x}{\bar{x}} \cdot 100\%$$


## Постановка задачи

Для заданных выборочных данных вычислить с использованием метода моментов и условных вариант точечные статистические оценки математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса, моды, медианы и коэффициента вариации исследуемой случайной величины. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы


In [1]:
import pandas as pd
import math
import matplotlib.pyplot as plt
import numpy as np

In [2]:
h_lenth = 3.59
h_depth = 1.14
k = 7
N = 117

In [3]:
df_lenth = pd.read_json('df_stats_len.json')
display(df_lenth)

,Интервал,Середина интервала,Абсолютная частота (m_i),Относительная частота (m_i/N),Накопленная абс. частота,Накопленная отн. частота
1,"[34.5, 37.9]",36.20,17,0.1453,17,0.1453
2,"[38.1, 41.6]",39.85,23,0.1966,40,0.3419
3,"[42.0, 45.2]",43.60,23,0.1966,63,0.5385
4,"[45.3, 48.7]",47.00,35,0.2991,98,0.8376
5,"[49.0, 50.7]",49.85,17,0.1453,115,0.9829
6,[54.3],54.30,1,0.0085,116,0.9915
7,[59.6],59.60,1,0.0085,117,1.0000


In [4]:
df_depth = pd.read_json('df_stats_depth.json')
display(df_depth)

,Интервал,Середина интервала,Абсолютная частота (m_i),Относительная частота (m_i/N),Накопленная абс. частота,Накопленная отн. частота
1,"[13.1, 14.2]",13.65,23,0.1966,23,0.1966
2,"[14.3, 15.3]",14.80,31,0.2650,54,0.4615
3,"[15.4, 16.5]",15.95,19,0.1624,73,0.6239
4,"[16.6, 17.5]",17.05,11,0.0940,84,0.7179
5,"[17.7, 18.8]",18.25,17,0.1453,101,0.8632
6,"[18.9, 19.5]",19.20,10,0.0855,111,0.9487
7,"[20.0, 21.1]",20.55,6,0.0513,117,1.0000


1. Для середин интервального ряда, полученного в практической работе №1, вычислим условные варианты

In [5]:
def conditional_options(df, h, c_i):
  data_rows = []
  C = df.loc[c_i]['Середина интервала']
  for i in range(1, k+1):
    x_i = df.loc[i]['Середина интервала']
    n_i = df.loc[i]['Абсолютная частота (m_i)']
    u_i = round((x_i - C) / h)
    data_rows.append({
        'i': i,
        'x_i': x_i,
        'n_i': n_i,
        'u_i': u_i,
        'n_i*u_i': n_i * u_i,
        'n_i*u_i^2': n_i * u_i**2,
        'n_i*u_i^3': n_i * u_i**3,
        'n_i*u_i^4': n_i * u_i**4,
        'n_i(u_i+1)^4': n_i * (u_i+1)**4
        })
  df = pd.DataFrame(data_rows, index=range(1, len(data_rows)+1))
  df.loc[len(df) + 1] = [
      'Σ',
      '-',
      round(df['n_i'].sum()),
      '—',
      df['n_i*u_i'].sum(),
      df['n_i*u_i^2'].sum(),
      df['n_i*u_i^3'].sum(),
      df['n_i*u_i^4'].sum(),
      df['n_i(u_i+1)^4'].sum(),
  ]
  display(df)
  print('∑ n_i(u_i+1)^4 = ', round(df.loc[len(df)]['n_i(u_i+1)^4'], 4))
  print('∑ n_i*u_i^4 + 4∑ n_i*u_i^3 + 6∑ n_i*u_i^2 + 4∑ n_i*u_i + N = ', end='')
  print(round(df.loc[len(df)]['n_i*u_i^4'] + 4*df.loc[len(df)]['n_i*u_i^3'] +
              6*df.loc[len(df)]['n_i*u_i^2'] + 4*df.loc[len(df)]['n_i*u_i']+N, 4))
  return df

Условный нуль для culmen_length_mm C = 47.0 (середина интервала с максимальной частотой)

In [6]:
print('Таблица №1 для признака culmen_length_mm')
co_len = conditional_options(df_lenth, h_lenth, 4)

Таблица №1 для признака culmen_length_mm


,i,x_i,n_i,u_i,n_i*u_i,n_i*u_i^2,n_i*u_i^3,n_i*u_i^4,n_i(u_i+1)^4
1,1,36.2,17,-3,-51,153,-459,1377,272
2,2,39.85,23,-2,-46,92,-184,368,23
3,3,43.6,23,-1,-23,23,-23,23,0
4,4,47.0,35,0,0,0,0,0,35
5,5,49.85,17,1,17,17,17,17,272
6,6,54.3,1,2,2,4,8,16,81
7,7,59.6,1,4,4,16,64,256,625
8,Σ,-,117,—,-97,305,-577,2057,1308


∑ n_i(u_i+1)^4 =  1308
∑ n_i*u_i^4 + 4∑ n_i*u_i^3 + 6∑ n_i*u_i^2 + 4∑ n_i*u_i + N = 1308


Равенство выполнено, следовательно вычисления произведены верно.

Условный нуль для culmen_depth_mm C = 14.8

In [7]:
print('\nТаблица №1 для признака culmen_depth_mm')
co_depth = conditional_options(df_depth, h_depth, 2)


Таблица №1 для признака culmen_depth_mm


,i,x_i,n_i,u_i,n_i*u_i,n_i*u_i^2,n_i*u_i^3,n_i*u_i^4,n_i(u_i+1)^4
1,1,13.65,23,-1,-23,23,-23,23,0
2,2,14.8,31,0,0,0,0,0,31
3,3,15.95,19,1,19,19,19,19,304
4,4,17.05,11,2,22,44,88,176,891
5,5,18.25,17,3,51,153,459,1377,4352
6,6,19.2,10,4,40,160,640,2560,6250
7,7,20.55,6,5,30,150,750,3750,7776
8,Σ,-,117,—,139,549,1933,7905,19604


∑ n_i(u_i+1)^4 =  19604
∑ n_i*u_i^4 + 4∑ n_i*u_i^3 + 6∑ n_i*u_i^2 + 4∑ n_i*u_i + N = 19604


Равенство выполнено, следовательно вычисления произведены верно.

2. Вычислим условные эмпирические моменты $\bar{M}^*_k$ через условные варианты. С помощью условных эмпирических моментов вычислим центральные эмпирические моменты $\bar{m}_k$.

In [8]:
def moments(df, h, c_i):
  data_rows = []
  C = df.loc[c_i]['x_i']
  size = len(df)
  M_1 = round(df.loc[size]['n_i*u_i']/N, 2)
  M_2 =  round(df.loc[size]['n_i*u_i^2']/N, 2)
  M_3 =  round(df.loc[size]['n_i*u_i^3']/N, 2)
  M_4 =  round(df.loc[size]['n_i*u_i^4']/N, 2)
  M = [M_1, M_2, M_3, M_4]
  m_1 = 0
  m_2 = round((M_2 - M_1**2) * h**2, 2)
  m_3 = round((M_3 - 3*M_2*M_1 + 2*M_1**3) * h**3, 2)
  m_4 = round((M_4 - 4*M_3*M_1 + 6*M_2*M_1**2 - 3*M_1**4) * h**4, 2)
  m = [m_1, m_2, m_3, m_4]
  for i in range(1, 5):
    data_rows.append({
        'k': i,
        'M̄*_k': M[i-1],
        'm̄_k': m[i-1]
    })
  df = pd.DataFrame(data_rows, index=range(1, len(data_rows)+1))
  display(df)
  print('Выборочные среднее x_в = ', round(M_1 * h + C, 2))
  print('СКО = ', round(m_2**(1/2), 2))
  return df

In [9]:
print('Для признака длина клюва:')
moments_len = moments(co_len, h_lenth, 4)

Для признака длина клюва:


,k,M̄*_k,m̄_k
1,1,-0.83,0.00
2,2,2.61,24.76
3,3,-4.93,19.68
4,4,17.58,1756.84


Выборочные среднее x_в =  44.02
СКО =  4.98


In [10]:
print('Для признака высота клюва:')
moments_depth = moments(co_depth, h_depth, 2)

Для признака высота клюва:


,k,M̄*_k,m̄_k
1,1,1.19,0.00
2,2,4.69,4.25
3,3,16.52,4.66
4,4,67.56,38.44


Выборочные среднее x_в =  16.16
СКО =  2.06


3. Вычислим выборочные среднее $\bar{x}_в$ и дисперсию $D_в$ с помощью стандартной формулы и с помощью условных вариант. Вычислим выборочное СКО
$σ_в$

In [11]:
def calculate_x_D(df):
  product  = []
  squared_deviations = []
  for idx, row in df.iterrows():
      x_j = row['Середина интервала']
      n_j = row['Относительная частота (m_i/N)']
      product.append(x_j * n_j)

  x_B = round(sum(product), 2)
  for idx, row in df.iterrows():
    x_j = row['Середина интервала']
    n_j = row['Относительная частота (m_i/N)']
    squared_deviations.append((x_j - x_B) ** 2 * n_j)
  D = round(sum(squared_deviations), 2)
  sigma = round(D**(0.5), 2)
  print(x_B, D, sigma)
  return x_B, D, sigma

In [12]:
print('Выборочные среднее x¯в, дисперсия Dв и СКО σв\n')
print('C помощью стандартной формулы:')
print('Для признака длина клюва:')
x_B_l, D_l, sigma_l = calculate_x_D(df_lenth)
print('Для признака высота клюва:')
x_B_d, D_d, sigma_d = calculate_x_D(df_depth)

Выборочные среднее x¯в, дисперсия Dв и СКО σв

C помощью стандартной формулы:
Для признака длина клюва:
43.94 22.89 4.78
Для признака высота клюва:
16.15 4.22 2.05


С помощью стандартной формулы и с помощью условных вариант соответственно:

Для длины клюва:

выборочные среднее $\bar{x}_в$ = 43.94 и 44.02


оценка дисперсии $D_в$ = 22,89 и 24.76

СКО $\sigma$ = 4,78

Для высоты клюва:

выборочные среднее $\bar{x}_в$ = 16.15 и 16.16

оценка дисперсии $D_в$ = 4,22 и 4.25

 СКО $\sigma$ = 2,05

Результаты получились достаточно близкими

4. Вычислим исправленную выборочную дисперсию $s_2$ и исправленное СКО $s$.

In [13]:
s_2_l = round(N/(N-1) * D_l, 2)
s_l = round(s_2_l**(0.5), 2)
print(s_2_l, s_l)

23.09 4.81


In [14]:
s_2_d = round(N/(N-1) * D_d, 2)
s_d = round(s_2_d**(0.5), 2)
print(s_2_d, s_d)

4.26 2.06


Для длины клюва:

оценка дисперсии $D_в$ = 22,91 и СКО $\sigma$ = 4,79

исправленная оценка дисперсии $s^2$ = 23,11 и СКО s = 4,81


Для высоты клюва:

 оценка дисперсии $D_в$ = 4,22 и СКО $\sigma$ = 2,05

исправленная оценка дисперсии $s^2$ = 4,26 и СКО s = 2,06

Результаты для смещенной и исправленной оценки получились достаточно близкими

5. Найдем статистическую оценку коэффициентов асимметрии
$\bar{A}_s$ и эксцесса $\bar{E}_в$

In [15]:
def calculate_A_E(df, h, c_i, D):
  data_rows = []
  C = df.loc[c_i]['x_i']
  size = len(df)
  M_1 = round(df.loc[size]['n_i*u_i'] / N, 2)
  M_2 =  round(df.loc[size]['n_i*u_i^2'] / N, 2)
  M_3 =  round(df.loc[size]['n_i*u_i^3'] / N, 2)
  M_4 =  round(df.loc[size]['n_i*u_i^4'] / N, 2)
  M = [M_1, M_2, M_3, M_4]
  m_1 = round(M_1 * h + C, 2)
  m_2 = round((M_2 - M_1**2) * h**2, 2)
  m_3 = round((M_3 - 3*M_2*M_1 + 2*M_1**3) * h**3, 2)
  m_4 = round((M_4 - 4*M_3*M_1 + 6*M_2*M_1**2 - 3*M_1**4) * h**4, 2)

  s_3 = ((N/(N-1) * D)**(0.5))**3
  s_4 = (N/(N-1) * D)**2

  A = round(m_3/s_3, 2)
  E = round(m_4/s_4 - 3, 2)
  return A, E

In [16]:
A_len, E_len = calculate_A_E(co_len, h_lenth, 4, D_l)
print("Статистические оценки коэффициентов асимметрии 𝐴¯𝑠 и эксцесса  𝐸¯в для длины клюва", "\nA = ", A_len,"\nE = ", E_len)

Статистические оценки коэффициентов асимметрии 𝐴¯𝑠 и эксцесса  𝐸¯в для длины клюва 
A =  0.18 
E =  0.3


Для длины клюва коэффициент ассиметрии больше 0, что означает, что в распределение умеренно скошено впарво.

Так как коэффициент эксцесса больше 0, то распределение островершинное относительно нормального распределения.

In [17]:
A_d, E_d = calculate_A_E(co_depth, h_depth, 2, D_d)
print("Статистические оценки коэффициентов асимметрии 𝐴¯𝑠 и эксцесса  𝐸¯в для высоты клюва", "\nA = ", A_d,"\nE = ", E_d)

Статистические оценки коэффициентов асимметрии 𝐴¯𝑠 и эксцесса  𝐸¯в для высоты клюва 
A =  0.53 
E =  -0.88


Для высоты клюва коэффициент ассиметрии больше 0, что говорит о том, что распределение скошено вправо и, так как значение больше 0.5, то ассиметрия существенная.

Коэффициент эксцесса меньше 0, значит распределение является пологим.

6. Для интервального ряда вычислим моду, медиану и коэффициент вариации $V^∗$

Мода интервального ряда определяется через интервал с наибольшей частотой.

Для признака длина клюва это 4-ый интервал [45.3, 48.7] с относительной частотой 0,2991.

Для признака высота клюва это 2-ой интервал [14.3, 15.3] с относительной частотой 0,2650.

In [18]:
def mode(df, i, h):
  x_i = df.loc[i]['Интервал'][0]
  m_i = df.loc[i]['Относительная частота (m_i/N)']
  m_0 = df.loc[i-1]['Относительная частота (m_i/N)']
  m_1 = df.loc[i+1]['Относительная частота (m_i/N)']
  M_0 = x_i + h * (m_i - m_0) / (2*m_i - m_0 - m_1)
  return round(M_0, 4)

mode_length = mode(df_lenth, 4, h_lenth)
mode_depth =  mode(df_depth, 2, h_depth)
print('Мода для признака длина клюва = ', mode_length)
print('Мода для признака высота клюва = ', mode_depth)

Мода для признака длина клюва =  46.7357
Мода для признака высота клюва =  14.756


Значение медианы принадлежит интервалу, для которого накопленная относительная частота равна или превышает 0.5, а накопленная относительная частота для предыдущего интервала меньше 0,5.

Таким образом для признака длина клюва это 3-ый интервал [42.0, 45.2] с накопленной относительной частотой 0,5385.

А для  признака высота клюва это 3-ой интервал [15.4, 16.5] с накопленной относительной частотой 0,6239.

In [19]:
def median(df, i, h):
  x_i = df.loc[i]['Интервал'][0]
  m_i = df.loc[i]['Относительная частота (m_i/N)']
  sum_m_i_1 = df.loc[i-1]['Накопленная отн. частота']
  M_e = x_i + h / m_i * (0.5 - sum_m_i_1)
  return round(M_e, 4)

median_length = median(df_lenth, 3, h_lenth)
median_depth =  median(df_depth, 3, h_depth)
print('Мода для признака длина клюва = ', median_length)
print('Мода для признака высота клюва = ', median_depth)

Мода для признака длина клюва =  44.887
Мода для признака высота клюва =  15.6703


In [20]:
V_l = round(sigma_l/x_B_l*100, 2)
print('Выборочный коэффициент вариации для признака длина клюва = ', V_l)
V_d = round(sigma_d/x_B_d*100, 2)
print('Выборочный коэффициент вариации для признака высота клюва = ', V_d)

Выборочный коэффициент вариации для признака длина клюва =  10.88
Выборочный коэффициент вариации для признака высота клюва =  12.69


Оба признака характеризуются небольшой вариацией.

## Выводы

В ходе работы были вычислены условные варианты и проведен контроль вычислений. Вычислены условные эмпирические моменты $\bar{M}^*_k$ через условные варианты. С помощью условных эмпирических моментов вычислены центральные эмпирические моменты $\bar{m}_k$.

Найдены выборочные среднее $\bar{x}_в$ и дисперсию $D_в$ с помощью стандартной формулы и с помощью условных вариант. Для длины клюва получились следующие значения: $\bar{x}_в$ = 43.94, $D_в$ = 22.89, $\sigma$ = 4.78.
Для высоты клюва: $\bar{x}_в$ = 16.15, $D_в$ = 4.22, $\sigma$ = 2.05.

Также были рассчитаны исправленная выборочная дисперсия $s^2$ и исправленная СКО $s$. Для длины клюва получены значения: исправленная оценка дисперсии $s^2$ = 23,11 и СКО s = 4,81.
Для высоты клюва: исправленная оценка дисперсии $s^2$ = 4,26 и СКО s = 2,06.

Найдена статистическая оценка коэффициентов асимметрии
$\bar{A}_s$ и эксцесса $\bar{E}_в$. Распределение для длины клюва островершинно и умеренно скошено впарво. Для высоты клюва выявлена существенная ассиметрия, распределение скошено вправо, а также является пологим.

Для интервального ряда вычислены мода, медиана и коэффициент вариации $V^∗$. Для признака длина клюва мода - 4-ый интервал с относительной частотой 0,2991. Для признака высота клюва это 2-ой интервал с относительной частотой 0,2650.

Для длины клюва медиана это 3-ий интервал, для высоты клюва медиана также 3-ий интервал. Разброс значений вариант относительно выборочного среднего для характеристики длина клюва получился 10.9%, для признака высота клюва - 12.7%.
